sys/psa_crypto/psa_persistent_storage/psa_crypto_cbor_encoder.c

```diff
@@ -0,0 +1,625 @@
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+
+    return PSA_SUCCESS;
+
+error:
+    DEBUG("[persistent storage] error while encoding asymmetric key pair: %s",
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+}
+#endif /* PSA_ASYMMETRIC_KEYPAIR_COUNT */
+
+#if PSA_SINGLE_KEY_COUNT
+/**
+ * @brief   Encode contents of a PSA key slot containing a single key
+ *
+ * @param   enc     Active nanocbor encoder
+ * @param   slot    Pointer to key slot to encode
+ * @return  psa_status_t
+ */
+static psa_status_t psa_encode_single_key(nanocbor_encoder_t *enc, psa_key_slot_t *slot)
+{
+    int res;
+    size_t *key_data_len;
+    uint8_t *key_data;
+    int key_array_len = 2;
+    psa_get_key_data_from_key_slot(slot, &key_data, &key_data_len);
+
+    res = nanocbor_fmt_array(enc, key_array_len);
+    if (res < 0) {
+        DEBUG("[persistent storage] error while encoding single key slot: %s",
+                                                    nanocbor_error_to_string(res));
+        return nanocbor_error_to_psa_status(res);
+    }
+
+    return psa_encode_key_data(enc, key_data, *key_data_len);
+}
+#endif /* PSA_SINGLE_KEY_COUNT */
+
+/**
+ * @brief   Encode a set of key attributes in CBOR
+ *
+ * @param   enc     Active nanocbor encoder
+ * @param   attr    Pointer to key attributes to encode
+ * @return  psa_status_t
+ */
+static psa_status_t psa_encode_key_attributes(nanocbor_encoder_t *enc, psa_key_attributes_t *attr)
+{
+    int res;
+
+    res = nanocbor_fmt_array(enc, 5);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_fmt_uint(enc, attr->id);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_fmt_uint(enc, attr->type);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_fmt_uint(enc, attr->bits);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_fmt_uint(enc, attr->lifetime);
+    if (res < 0) {
+        goto error;
+    }
+
+    /* Policy */
+    res = nanocbor_fmt_array(enc, 2);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_fmt_uint(enc, attr->policy.usage);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_fmt_uint(enc, attr->policy.alg);
+    if (res < 0) {
+        goto error;
+    }
+
+    return PSA_SUCCESS;
+
+    error:
+        DEBUG("[persistent storage] error while encoding key attributes: %d = %s", res,
+                                                    nanocbor_error_to_string(res));
+        return nanocbor_error_to_psa_status(res);
+}
+
+psa_status_t psa_encode_key_slot(psa_key_slot_t *slot, uint8_t *output,
+                                     size_t output_len, size_t *output_size)
+{
+    int res;
+    psa_status_t status;
+    nanocbor_encoder_t enc;
+    nanocbor_encoder_init(&enc, output, output_len);
+    *output_size = 0;
+
+    /* Key Struct */
+    res = nanocbor_fmt_array_indefinite(&enc);
```

Comment:
Huu, don't we know the number of (immediate) elements in the array?

Reply:
Yep, we do, fixed it